clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -424,9 +424,59 @@ caseCase _ e = return e
 {-# SCC caseCase #-}
 
 -- | Inline function with a non-representable result if it's the subject
--- of a Case-decomposition
+-- of a Case-decomposition. It's a custom topdown traversal that -for efficiency
+-- reasons- does not explore alternative of cases whose subject triggered an
+-- 'inlineNonRepWorker'.
 inlineNonRep :: HasCallStack => NormRewrite
-inlineNonRep _ e@(Case scrut altsTy alts)
+inlineNonRep _ctx = go
+ where
+  go :: HasCallStack => Term -> NormalizeSession Term
+  go e0 = do
+    (e1, hasChanged) <- listen (inlineNonRepWorker e0)
+    if Monoid.getAny hasChanged then
+      pure e1
+    else
+      case e1 of
+        Var _ -> pure e1
+        Data _ -> pure e1
+        Literal _ -> pure e1
+        Prim _ -> pure e1
+        Lam v t -> Lam v <$> go t
+        TyLam v t -> TyLam v <$> go t
+        App a0 b0 -> App <$> go a0 <*> go b0
+        TyApp t typ -> TyApp <$> go t <*> pure typ
+        Cast t0 typ0 typ1 -> Cast <$> go t0 <*> pure typ0 <*> pure typ1
+        Tick tickInfo t -> Tick tickInfo <$> go t
+        Letrec bindings_ body0 -> do
+          let (bndrs, bodies0) = unzip bindings_
+          bodies1 <- mapM go bodies0
+          body1 <- go body0
+          pure (Letrec (zip bndrs bodies1) body1)
+        Case subj0 typ alts -> do
+          (subj1, subjHasChanged) <- listen (go subj0)
+          if Monoid.getAny subjHasChanged then
+            -- This is the crucial part of this transformation, making it
+            -- substantially different from just a bottom-up traversal with
+            -- 'inlineNonRepWorker': if any non representable binder is inlined
+            -- in the the subject of a case expression, we don't traverse the
+            -- alternatives as this inlined structure might very well give use a
+            -- chance to do 'caseCon'. Traversing the alternatives might
+            -- therefore yield a lot of work.
+            pure (Case subj1 typ alts)
```

Comment:
I think this needs to be `changed`, I'm not sure the change propagates when you capture it locally with `listen`

Reply:
It does:

```haskell
listens :: (Monoid w, Monad m) => (w -> b) -> RWST r w s m a -> RWST r w s m (a, b)
listens f m = RWST $ \ r s w -> do
    (a, s', w') <- runRWST m r s
    let wt = w `mappend` w'
    wt `seq` return ((a, f w'), s', wt)
{-# INLINE listens #-}
```

The `wt` is getting propagated.

Reply:
Oh and `listen = listens id`

Reply:
:+1: